dev/benchmarks/macrobenchmarks/lib/src/web/recorder.dart

```diff
@@ -332,175 +406,168 @@ class _WidgetBuildRecorderHostState extends State<_WidgetBuildRecorderHost> {
-      'name': name,
-      'scoreKeys': <String>['averageDrawFrameDuration'],
-      'averageDrawFrameDuration': averageDrawFrameDuration.inMicroseconds,
-      'drawFrameDurationNoise': drawFrameDurationNoise,
-      'frames': frames
-          .map((FrameMetrics frameMetrics) => frameMetrics.toJson())
-          .toList(),
-    };
+  /// Invokes [callback] and records the duration of its execution under [key].
+  Duration record(String key, VoidCallback callback) {
+    final Duration duration = timeAction(callback);
+    addDataPoint(key, duration);
+    return duration;
   }
 
-  @override
-  String toString() {
-    return _formatToStringLines(<String>[
-      'benchmark: $name',
-      'averageDrawFrameDuration: ${averageDrawFrameDuration.inMicroseconds}μs',
-      'drawFrameDurationNoise: ${drawFrameDurationNoise * 100}%',
-      'frames:',
-      ...frames.expand((FrameMetrics frame) =>
-          '$frame\n'.split('\n').map((String line) => '- $line\n')),
-    ]);
+  void addDataPoint(String key, Duration duration) {
+    scoreData
+        .putIfAbsent(key, () => Timeseries())
+        .add(duration.inMicroseconds);
   }
-}
 
-/// Contains metrics for a single frame.
-class FrameMetrics {
-  FrameMetrics._({
-    @required this.drawFrameDuration,
-    @required this.sceneBuildDuration,
-    @required this.windowRenderDuration,
-  });
+  /// Decides whether the data collected so far is sufficient to stop, or
+  /// whether the benchmark should continue collecting more data.
+  ///
+  /// The signals used are sample size, noise, and duration.
+  ///
+  /// If any of the timeseries doesn't satisfy the noise requirements, this
+  /// method will return true (asking the benchmark to continue collecting
+  /// data).
+  bool shouldContinue() {
+    // If we haven't recorded anything yet, we don't wanna stop now.
+    if (scoreData.isEmpty) {
+      return true;
+    }
+
+    final List<bool> shouldContinueList = <bool>[];
+    for (final String key in scoreData.keys) {
+      final Timeseries timeseries = scoreData[key];
 
-  /// Total amount of time taken by [Window.onDrawFrame].
-  final Duration drawFrameDuration;
+      // Collect enough data points before considering to stop.
+      if (timeseries.count < _kMinSampleCount) {
+        shouldContinueList.add(true);
+        continue;
+      }
+
+      // Is it still too noisy?
+      if (timeseries.noise > _kNoiseThreshold) {
+        // If the timeseries has enough data, stop it, even if it's noisy under
+        // the assumption that this benchmark is always noisy and there's nothing
+        // we can do about it.
+        if (timeseries.count > _kMaxSampleCount) {
+          print(
+            'WARNING: Noise of benchmark "$name.$key" did not converge below '
+            '${_percent(_kNoiseThreshold)}. Stopping because it reached the '
+            'maximum number of samples $_kMaxSampleCount. Noise level is '
+            '${_percent(timeseries.noise)}.',
+          );
```

Comment:
It would be misleading to print this message here because we may not be stopping yet. Other timeseries may still report `true`, in which case we'll print this but continue running the benchmark. Perhaps we should accumulate this info in a buffer and only print it after we determined that the benchmark should stop?